runner/src/simulate.rs

```diff
@@ -95,58 +117,327 @@ impl Scenario {
+                    let url = format!("http://{}:{}", host, metrics_port);
+                    let resp = client.get(&url).send().await?;
+                    if !resp.status().is_success() {
+                        warn!(?resp, "metrics request failed for peer {}", addr);
+                    } else {
+                        let body = resp.text().await?;
+                        let mut found = false;
+                        for metric in body.lines() {
+                            match metric.split(' ').collect::<Vec<&str>>().as_slice() {
+                                [name, value] if *name == metric_name => {
+                                    let val = value.parse::<u64>().map_err(|e| {warn!(metric=%name, %value, "failed to parse metric: {}", e); e}).ok();
+                                    results.push(val);
+                                    found = true;
+                                    break;
+                                }
+                                _ => {}
+                            }
+                        }
+                        if !found {
+                            warn!("Failed to find metric {} for host: {}", metric_name, addr);
+                            results.push(None);
+                        }
+                    }
+                } else {
+                    warn!("Failed to parse ceramic addr for host: {}", addr);
+                }
+            }
+        }
+        Ok(results)
+    }
+
+    async fn collect_before_metrics(&mut self) -> Result<()> {
+        if !self.manager {
+            Ok(())
+        } else {
+            match self.scenario {
+                Scenario::IpfsRpc
+                | Scenario::CeramicSimple
+                | Scenario::CeramicWriteOnly
+                | Scenario::CeramicNewStreams
+                | Scenario::CeramicQuery
+                | Scenario::CeramicModelReuse => Ok(()),
+                Scenario::EventIdSync => {
+                    let peers = self
+                        .get_peers_counter_metric(EVENT_SYNC_METRIC_NAME, IPFS_SERVICE_METRICS_PORT)
+                        .await?;
+                    let res: Vec<u64> = peers.iter().filter_map(|v| *v).collect();
+                    if res.len() != peers.len() {
+                        bail!(
+                            "Failed to collect metrics for all peers before scenario {:?}: {:?}",
+                            self.scenario,
+                            peers
+                        )
+                    }
+                    self.before_metrics = Some(res);
+                    Ok(())
+                }
+            }
+        }
+    }
+
+    /// For now, most scenarios are successful if they complete without error and only EventIdSync has a criteria.
+    /// Not a result to ensure we always proceed with cleanup, even if we fail to validate the scenario.
+    async fn validate_scenario_success(&self, metrics: &GooseMetrics) -> CommandResult {
```

Comment:
I started to refactor things a bit to make this more testable (e.g. inject a metrics collector trait) and write a couple tests. I'm planning to push that through in a follow up PR so we're not blocked by having the scenario.